### ISSUE TITLE
Why do you have the binaries for the dev-tools directly in the repo?

### ISSUE THREAD

**Issue by sypets** [FriendsOfTYPO3/tea#255]
and not let them install via composer and have them in a require-dev section? Which would also automatically get them updated when new versions are available and I think makes maintenance easier. (If you want a specific version you can set a specific version in the composer.json but I don't really see a need for that if semantic versioning is used).

e.g. `tools/` includes `phpcs`, `php-cs-fixer`, `phpstan` etc.

instead of having composer.json, e.g.:

```
"require-dev": {
                "friendsofphp/php-cs-fixer": "^2.18.6",
                "jangregor/phpstan-prophecy": "^0.8.1",
                "phpstan/phpstan": "^0.12.64",

etc. ...
```

----

This would also be more like it's done in the TYPO3 core, which makes it easier for people to get started,  contribute and transition from core contributions to extension contributions and vice versa, see my issue #241

**Comment by oliverklee** [comment]
These are actually two questions:

Q: Why are we installing the dev tools via PHIVE as PHARs instead of via Composer?

A: To avoid dependency hell, and also to be able to use tools that require a PHP version higher than the lowest allowed version for this extension (7.2).

Q: Why are we versioning the tool PHARs instead of using symlinks and/or `.gitignore`

A: No real reason. We'll need to discuss this.

**Comment by DanielSiepmann** [comment]
We've discussed the `.gitignore` topic at our be-weekly call.
You can also find info here: https://twitter.com/oliklee/status/1380082284200595457.

@oliverklee will add documentation about the how.

**Comment by sypets** [comment]
To have the binaries in the project makes it - to me - very untransparent and just "feels weird". What version is being used? Ok, I can checkout the repo and (usually) do a `script --version`. But if the version is in `composer.json` / `composer.lock` you can immediately see the version, you can use other tools to check, you can use the same version in other projects etc.

I get that if you use Composer to install you might have version conflicts. But, is this really an issue? For which tools do you expect version conflicts?

Is this a best practice, will people coming from other PHP projects feel right at home here?

My wish is that the decisions here are somehow documented and have gone through a review. There are a number of decisions you made here and you have put a great deal of though into it but for me - as a non expert and still newbie in some things - it is not something I can follow and understand. (I can follow and understand most of the explanation here but some answers raise further questions ...).

Unfortunately everyone is doing it differently and this very repo tries to resolve that. But I think you need some documentation of your thought process and your decisions - otherwise these questions will come up again.

For example:

1. Supporting several TYPO3 versions at the same time in one extension, pros / cons
2. use binaries in repo vs. install via composer
3. etc. 

See also my stack overflow question https://stackoverflow.com/questions/64713267/strategies-for-supporting-multiple-typo3-versions-in-extensions

That is still not really resolved for me. In my extension, I started out supporting 9 and 10. But this made the maintaining difficult - especially if you start adding more tools and use the extensions scanner etc. So I eventually dropped 9 support or rather moved it to a branch.

I don't know if these topics have already been discussed at length - I think these (and other topics) are very important - to make it easier to maintain TYPO3 extensions. For me, there are a lot of open questions, a lot of things not documented or difficult to figure out how to do. 

Just copy-pasting stuff from this extension feels weird. I don't really want to do that unless I understand the what and why and if this solution is really a good fit for me.

**Comment by oliverklee** [comment]
> What version is being used?

You can see that in the file [phive.xml](https://github.com/TYPO3-Documentation/tea/blob/main/phive.xml), which works a bit like mixture of `composer.json` and `composer.lock`.

> For which tools do you expect version conflicts?

Mostly for PHPStan and TypoScript-lint.

Generally, I recommend having all stand-alone tools as PHARs via PHIVE, and only have those via Composer that cannot be installed via PHIVE, instead of installing everything via Composer, and move things to PHIVE when there is a problem.

There is another benefit: We can have tools via  PHIVE in versions that do not work with all PHP versions supported, which does not work with Composer.

Yes, we need to document all this. I have added #264 and #265 for this. Thanks for pointing this out!

**Comment by oliverklee** [comment]
Oh, and concerning having the binaries in the repo: I'd like to switch to installing the PHARS via PHIVE during the CI and installation process instead. (This didn't work until now due to a bug in PHIVE which made non-interactive installations near impossible. When a new version of PHIVE is released, we should be able to switch.)

**Comment by oliverklee** [comment]
Fixed by #1277.
